### DISCUSSION TITLE
How to test a formula

### DISCUSSION THREAD

**Discussion by alexreg** [Homebrew/discussions#4631] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.28-21-g9708a57
ORIGIN: https://github.com/Homebrew/brew
HEAD: 9708a57935a538d133093d63aaf74028d023b116
Last commit: 5 minutes ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 72a0664d6f7058274d125c23a3f1d423ee306ded
Core tap last commit: 25 minutes ago
Core tap branch: bazarr
Core tap JSON: 23 Jun 15:33 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_BROWSER: open -a Safari.app
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.fkMFCm4aa8/org.xquartz:0
HOMEBREW_EDITOR: mate_wait
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.41.0 => /opt/homebrew/bin/git
Curl: 7.88.1 => /usr/bin/curl
macOS: 13.4-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: 14.3.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  subliminal

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/BandedSmithWaterman.h
  /usr/local/include/BedReader.h
  /usr/local/include/Fasta.h
  /usr/local/include/IndelAllele.h
  /usr/local/include/IntervalTree.h
  /usr/local/include/LargeFileSupport.h
  /usr/local/include/LeftAlign.h
  /usr/local/include/Mosaik.h
  /usr/local/include/Region.h
  /usr/local/include/Repeats.h
  /usr/local/include/SmithWatermanGotoh.h
  /usr/local/include/Variant.h
  /usr/local/include/disorder.h
  /usr/local/include/filevercmp.h
  /usr/local/include/join.h
  /usr/local/include/makeUnique.h
  /usr/local/include/mt19937ar.h
  /usr/local/include/multichoose.h
  /usr/local/include/multipermute.h
  /usr/local/include/vec128int.h
  /usr/local/include/veclib_types.h

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  rust
  dotnet

Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
  git -C $(brew --repo homebrew/core) checkout master
```

### Description of issue

I am writing an update for a formula (bazarr) on a branch of homebrew-core forked from master. When I run `brew install -s bazarr`, it uses some online version of the formula rather than the version I'm working on the branch. How can I avoid this? (This seems to be somewhat new behaviour that I didn't use to experience.)
*Excerpt of output*
```
...
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-core/ef91fcd1ab07718738f59b845cdb41b2b745cdf5/Formula/bazarr.rb
...
```

**Comment by Bo98** [comment]
You can do `HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr`. Maybe we should add a flag for it instead - I'll take a look.

**Reply by alexreg** [reply]
Thank you. A flag would be nice (perhaps implied by `-s` but not necessarily). This will do in the meanwhile.

**Reply by alexreg** [reply]
Actually, I just tested that, and unfortunately it has no effect. It still downloads the formula, rather than using my local version. Odd.

**Reply by osalbahr** [reply]
Btw, what do you see when you run `brew developer`? I think assuming `-s` when a formula is modified and developer mode is on would be a good idea.

**Reply by osalbahr** [reply]
> Actually, I just tested that, and unfortunately it has no effect. It still downloads the formula, rather than using my local version. Odd.

What do you see when you `cd $(brew --repo homebrew/core) && git status`?

FYI I was able to reproduce the unexpected behavior in `podman -it homebrew/brew` of `-s` not always installing from source. But HOMEBREW_NO_INSTALL_FROM_API=1` did the trick:
```console
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   Formula/bazarr.rb

no changes added to commit (use "git add" and/or "git commit -a")
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core$ cat Formula/bazarr.rb
paste the 
content of
the file here
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core$ cd $(brew --repo)
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew$ git checkout master
Already on 'master'
Your branch is up to date with 'origin/master'.
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew$ brew update
Already up-to-date.
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew$ HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr
Error: bazarr: undefined local variable or method `the' for Formulary::FormulaNamespacef2c335cd346b688644c8c15740a76670:Module
Did you mean?  then
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew$ brew install -s bazarr
==> Fetching dependencies for bazarr: aom, zlib, libpng, aribb24, dav1d, bzip2, freetype, ncurses, readline, libxcrypt, util-linux, expat, fontconfig, frei0r, ca-certificates, gmp, libunistring, libidn2, libtasn1, nettle, libffi, p11-kit, openssl@3, libevent, libnghttp2, unbound, gnutls, lame, fribidi, pcre2, dbus, mpdecimal, sqlite, xz, berkeley-db@5, libedit, krb5, libtirpc, libnsl, unzip, python@3.11, glib, xorgproto, libxau, libxdmcp, libxcb, libx11, libxext, libxrender, lzo, pixman, cairo, graphite2, icu4c, harfbuzz, libunibreak, libass, libxml2, libbluray, cjson, mbedtls, librist, libsoxr, libvidstab, libvmaf, libogg, libvorbis, libvpx, opencore-amr, jpeg-turbo, lz4, zstd, libtiff, little-cms2, openjpeg, opus, rav1e, libsamplerate, flac, mpg123, libsndfile, isl, mpfr, libmpc, binutils, gcc, hwloc, open-mpi, fftw, ladspa-sdk, vamp-plugin-sdk, rubberband, libice, libxfixes, libxcursor, libxscrnsaver, libxxf86vm, json-c, m4, libtool, speexdsp, libcap, pulseaudio, libxi, libxinerama, libxrandr, xinput, sdl2, snappy, speex, srt, svt-av1, giflib, webp, leptonica, libb2, libarchive, pango, tesseract, theora, x264, x265, xvid, libsodium, zeromq, zimg, alsa-lib, libxv, ffmpeg, openblas, numpy, libimagequant, libraqm, tcl-tk, pillow, libobjc2, libgpg-error, libgcrypt, libxslt, gnustep-base, wavpack, unar, brotli, c-ares, libuv and node
```

@alexreg does following the steps in https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request#create-your-pull-request-from-a-new-branch make `HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr` install from source?

I also recommend running `"$(brew --repository homebrew/core)" && git status` because I found Homebrew to sometimes stash my changes (which I think is expected behavior) and I have to `git checkout master` or `brew developer on` to prevent it from happening. For the issue you're having, though, I don't think you have to run `brew developer on` and `git checkout master` should be enough. But this is unrelated to the `-s` bug.

```console
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core$ brew developer off
linuxbrew@b07646af8307:~/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core$ HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr
Error: bazarr: undefined local variable or method `the' for Formulary::FormulaNamespacef2c335cd346b688644c8c15740a76670:Module
Did you mean?  then
```

**Comment by alexreg** [comment]
Thanks for the tips. Still no luck I'm afraid. Here's the output of the commands you mentioned.

```sh
$ brew developer
Developer mode is enabled.
```

```sh
$ cd $(brew --repo homebrew/core) && git status
On branch bazarr
nothing to commit, working tree clean
```

Also, I agree about assuming `-s` (and `HOMEBREW_NO_INSTALL_FROM_API=1`) when a formula is modified and developer mode is on.

**Reply by osalbahr** [reply]
I think now you should be able to add the edit back and then `HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr`.

**Reply by alexreg** [reply]
How do you mean? The edit is already there, committed to the `bazarr` branch.

**Reply by osalbahr** [reply]
Oh, I see now. You’re locally in the `bazarr` branch. I thought you were on `master` for some reason.

So even though `bazarr.rb` is edited locally, developer mode is on, `HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr` doesn’t build locally? That’s odd.

You can test the changes in the `homebrew/core` Docker image (such as by running `podman run -it homebrew/brew`) but that’s a bit of an overkill.

**Reply by alexreg** [reply]
Yes, very odd... I'm not sure what's up.

**Reply by osalbahr** [reply]
Ok try to `cat` the formula file and see if the content is correct. If you send the file’s url in your repo I can try to apply and test your changes locally.

**Reply by alexreg** [reply]
Confirmed it's correct. Here's the file in my forked repo:
https://github.com/alexreg/homebrew-core/blob/bazarr/Formula/bazarr.rb

**Reply by osalbahr** [reply]
This is inside of the `podman run -it homebrew/brew` shell:
```console
...
linuxbrew@8dfc2b40a125:~$ HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr
bazarr 1.2.2 is already installed but outdated (so it will be upgraded).
==> Fetching bazarr
==> Downloading https://files.pythonhosted.org/packages/06/5a/e11cad7b79f2cf3dd2ff8f81fa8ca667e7591d3d8451768589996b65de
Already downloaded: /home/linuxbrew/.cache/Homebrew/downloads/0c3fb629c2e6acab76f090b78219aa2557b5fa6fcc55ff6399dd18746703c5a3--lxml-4.9.2.tar.gz
==> Downloading https://files.pythonhosted.org/packages/59/d9/17fe64f981a2d33c6e95e115c29e8b6bd036c2a0f90323585f1af639d5
Already downloaded: /home/linuxbrew/.cache/Homebrew/downloads/63531da8fecc2c6454bead647d65776b7e68cde9e712a72dae9c902f3144ce39--webrtcvad-wheels-2.0.11.post1.tar.gz
==> Downloading https://github.com/morpheus65535/bazarr/releases/download/v1.2.2/bazarr.zip
Already downloaded: /home/linuxbrew/.cache/Homebrew/downloads/14e1aa95e4c8f0881fa7c9ca0826a45a278b1e0a63d98334ade5049eab539ff8--bazarr.zip
==> Upgrading bazarr
  1.2.2 -> 1.2.2_1 

==> python3.11 -m venv --system-site-packages /home/linuxbrew/.linuxbrew/Cellar/bazarr/1.2.2_1/libexec
==> /home/linuxbrew/.linuxbrew/Cellar/bazarr/1.2.2_1/libexec/bin/pip install -v --no-deps --no-binary :all: --use-featur
==> /home/linuxbrew/.linuxbrew/Cellar/bazarr/1.2.2_1/libexec/bin/pip install -v --no-deps --no-binary :all: --use-featur
Error: An exception occurred within a child process:
  Errno::ENOENT: No such file or directory @ apply2files - /home/linuxbrew/.linuxbrew/Cellar/bazarr/1.2.2_1/libexec/data
...
linuxbrew@8dfc2b40a125:~$ history
    1  brew tap --force homebrew/core
    2  cd "$(brew --repository homebrew/core)"
    3  git remote add alexreg https://github.com/alexreg/homebrew-core.git
    4  git checkout master
    5  brew update
    6  git checkout -b temp origin/master
    7  brew edit bazarr
    8  curl https://raw.githubusercontent.com/alexreg/homebrew-core/bazarr/Formula/bazarr.rb > /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/bazarr.rb
    9  git status
   10  brew install bazarr
   11  bazarr --version
   12  bazarr --help
   13  brew test bazaar
   14  brew test bazarr
   15  brew install bazarr
   16  brew test bazarr
   17  brew install -s bazarr
   18  HOMEBREW_NO_INSTALL_FROM_API=1 brew install bazarr
   19  HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr
   20  cd
   21  HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr
   22  HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr --verbose
   23  HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr --debug
   24  HOMEBREW_NO_INSTALL_FROM_API=1 brew install -s bazarr --verbose --debug
   25  history
```

[bazar.log](https://github.com/Homebrew/discussions/files/12008773/bazar.log)

Does this help?

**Reply by alexreg** [reply]
@osalbahr Yes that does, thank you. Of course, I really would still like to get testing working locally though...

**Reply by alexreg** [reply, marked as answer]
Ahh, I see what the issue is: I use [brewdo](https://github.com/brewdo/brewdo) and this internally does a sudo to the `_homebrew` user. I needed to edit my sudoers file, and now it works. Thanks for bearing with me. Maybe we can still make this behaviour more automatic though.

**Reply by osalbahr** [reply]
> @osalbahr Yes that does, thank you. Of course, I really would still like to get testing working locally though...

I highly recommend installing the podman-desktop cask. After opening the app and waiting for the setup, you should be able to directly just `podman run -it homebrew/brew` and you will get the same outcome by typing the same commands in the container. Again, not ideal, and an overkill. Just an option. I’m not sure why following the contribution docs (the command I ran, with the suggested workaround added) doesn’t make you able to install from source.

**Reply by osalbahr** [reply]
> Ahh, I see what the issue is: I use [brewdo](https://github.com/brewdo/brewdo) and this internally does a sudo to the _homebrew user. I needed to edit my sudoers file, and now it works. Thanks for bearing with me. Maybe we can still make this behaviour more automatic though.

Oh. Glad you figured it out! If it’s not provided by Homebrew not sure what the solution might be for your use case. I never used brewdo (or even heard of it). Maybe upstream the automation solution to brewdo? Also, out of curiosity, why do you need a wrapper around Homebrew?

**Reply by osalbahr** [reply]
I just realized that `brew edit` gives a warning that the environment variable is needed to prevent ignoring the local formula:
```console
$ brew edit cbonsai
Warning: Unless `HOMEBREW_NO_INSTALL_FROM_API` is set when running
`brew install`, it will ignore your locally edited formula.
Editing /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cbonsai.rb
```

However, I don't think install from API should take precedent over `-s` (`--build-from-source`). Or alternatively updating the documentation in https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request#create-your-pull-request-from-a-new-branch.

**Reply by alexreg** [reply]
@osalbahr I use brewdo because I'm on a multi-user system, so it makes a lot of sense there. I think I'll just include a note in the brewdo installer to edit the sudoers file accordingly. Re your second comment, I agree that `-s` should imply `HOMEBREW_NO_INSTALL_FROM_API=1`, as this would mean considerably less surprising behaviour in general.

**Reply by osalbahr** [reply]
>  I agree that -s should imply HOMEBREW_NO_INSTALL_FROM_API=1, as this would mean considerably less surprising behaviour in general

Actually, I like your idea more than "take precedent" because it seems to be easier to implement and maintain. I was thinking along the lines of having `brew install` checking `-s` before checking the api flags, but having `-s` set the api flag seems easier

**Reply by alexreg** [reply]
@osalbahr Sounds good. :-)
